code/src/modules/object_tracker.cpp

```diff
@@ -143,45 +158,144 @@ void ObjectTracker::Run(FlightController *fc, void *opts) {
+    TIME_TYPE loop_start = last_loop;
+    //TIME_TYPE last_fix = sample_time - seconds(2);   //no fix (deprecate)
     bool had_fix = false;
 
+    Observation theGround = AssumptionGroundLevel();
+
+    Mat observation_map(observation_image_rows, observation_image_cols, CV_8UC4);
+
     while (!fc->CheckForStop()) {
+        last_loop = loop_start;
+        loop_start = steady_clock::now() - m_task_start;
+        TIME_TYPE loop_period = (loop_start - last_loop);
+
+        //clear the printable map
+        observation_map = Mat::zeros(observation_map.rows, observation_map.cols, CV_8UC4);
+
         double update_rate = 1.0 / fc->cam->GetFramerate();
-        auto sleep_time = microseconds((int)(1000000*update_rate));
+        TIME_TYPE sleep_time = microseconds((int)(1000000*update_rate));    //how long to wait for the next frame (FIXME)
+
 
         fc->cam->GetDetectedObjects(&locations);
         fc->fb->GetGimbalPose(&gimbal);
         fc->gps->GetLatest(&gps_position);
         fc->imu->GetLatest(&imu_data);
-        if (fc->lidar) {
-            lidar_distance = fc->lidar->GetLatest() / 100.0;
-        }
-        //Do we have an object?
-        if (locations.size() > 0) {                                             
+
+        double lidar_range = (double)fc->lidar->GetLatest() / (100.0); //convert lidar range to metres
+
+        
+        //Now would be a good time to use the velocity and acceleration handler
+        for(uint i=0; i<knownThings.size(); i++){
+            knownThings.at(i).updateObject(loop_period);
+
+            if(print_observation_map){  //plot the objects on the map
+                Distrib tmp = knownThings[i].getLocation();
+                rasterDistrib(&observation_map, &tmp, Vec4b(0,0,UCHAR_MAX,UCHAR_MAX), 1.0);  //red
+            }
+        }//blur the location.
+
+
+        //start making observation structures
+        Observation lidarObservation = ObservationFromLidar(loop_start, &gps_position, &gimbal, &imu_data, lidar_range);
+        rasterDistrib(&observation_map, &lidarObservation.location, Vec4b(UCHAR_MAX,0,0,UCHAR_MAX), 1.0);  //blue
+
+
+        //Did the camera see anything?
+        if (locations.size() > 0) {
             detected_object = locations.front();
-            SetCurrentState(fc, STATE_TRACKING_LOCKED);
+
+            //build observations for each
+            std::vector<Observation> visibles; //things we can currently see
+            visibles.reserve(locations.size()); //save multiple reallocations
+            for(uint i=0; i<locations.size(); i++){
+                visibles.push_back(ObservationFromImageCoords(loop_start, &gps_position, &gimbal, &imu_data, &detected_object));
+            }
+
+            if(print_observation_map){
+                //print the observations and objects
+                std::string filename = "tracker"+(static_cast<std::ostringstream*>( &(std::ostringstream() << observation_map_count++) )->str())+".png";
```

Comment:
Can be replaced with:
std::string filename = "tracker" + std::to_string(observation_map_count++) + ".png"